src/app/src/main/java/com/example/bookmark/SignupActivity.java

```diff
@@ -91,13 +90,15 @@ protected void onCreate(Bundle savedInstanceState) {
 
                         // add user and go back to main activity
                         User newUser = new User(username, firstName, lastName, emailAddress, phoneNumber);
-                        addUser(newUser, firebaseProvider);
+                        firebaseProvider.storeUser(newUser,
+                            aVoid -> {
+                            },
+                            e -> Toast.makeText(SignupActivity.this, "Connection Error. Please Try again.", Toast.LENGTH_LONG).show());
                         Intent intent = new Intent(SignupActivity.this, MainActivity.class);
                         intent.putExtra("SIGNED_UP_USERNAME", username);
                         startActivity(intent);
```

Comment:
The race condition still is possible as written. To clarify what my previous review wanted you to do:
```suggestion
                        OnSuccessListener<Void> onSuccessListener = aVoid -> {
                            Intent intent = new Intent(this, MainActivity.class);
                            intent.putExtra("SIGNED_UP_USERNAME", username);
                            startActivity(intent);
                        };
                        firebaseProvider.storeUser(newUser,
                            onSuccessListener,
                            e -> Toast.makeText(SignupActivity.this, "Connection Error. Please Try again.", Toast.LENGTH_LONG).show());
```